// @clr/angular/esm2015/utils/enums/key-codes.enum.js
/*
 * Copyright (c) 2016-2021 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
export var KeyCodes;
(function (KeyCodes) {
    KeyCodes["ArrowLeft"] = "ArrowLeft";
    KeyCodes["ArrowUp"] = "ArrowUp";
    KeyCodes["ArrowRight"] = "ArrowRight";
    KeyCodes["ArrowDown"] = "ArrowDown";
    KeyCodes["Backspace"] = "Backspace";
    KeyCodes["Tab"] = "Tab";
    KeyCodes["Enter"] = "Enter";
    KeyCodes["Escape"] = "Escape";
    KeyCodes["Space"] = "Space";
    KeyCodes["Home"] = "Home";
    KeyCodes["End"] = "End";
})(KeyCodes || (KeyCodes = {}));
export var IEKeyCodes;
(function (IEKeyCodes) {
    IEKeyCodes["ArrowUp"] = "Up";
    IEKeyCodes["ArrowDown"] = "Down";
    IEKeyCodes["ArrowRight"] = "Right";
    IEKeyCodes["ArrowLeft"] = "Left";
    IEKeyCodes["Space"] = "Spacebar";
    IEKeyCodes["Escape"] = "Esc";
    IEKeyCodes["Home"] = "Home";
    IEKeyCodes["End"] = "End";
    IEKeyCodes["Enter"] = "Enter";
    IEKeyCodes["Tab"] = "Tab";
})(IEKeyCodes || (IEKeyCodes = {}));
